src/work/work_trsmA.cc

```diff
@@ -90,38 +101,91 @@ void trsmA(Side side, scalar_t alpha, TriangularMatrix<scalar_t> A,
                         for (int64_t j = 0; j < nt; ++j) {
                             if (B.tileIsLocal(i, j)) {
-                                tile::scale( alpha, B(i, j) );
+                                if (target == Target::Devices) {
+                                    int device = B.tileDevice( i, j );
+
+                                    B.tileGetForWriting( i, j, device,
+                                            LayoutConvert( layout ) );
+
+                                    blas::Queue* queue = A.compute_queue( device, queue_0 );
+                                    assert( queue != nullptr );
+                                    auto T = B( i, j, device );
+                                    if (T.op() == Op::Trans || T.op() == Op::ConjTrans)
+                                        T = transpose( T );
+
+                                    device::gescale( T.mb(), T.nb(), alpha, one,
+                                            T.data(), T.stride(), *queue );
+                                    queue->sync();
+                                }
+                                else {
+                                    B.tileGetForWriting( i, j,
+                                            LayoutConvert( layout ) );
+                                    tile::scale( alpha, B(i, j) );
+                                }
                             }
                         }
                     }
                 }
 
                 // Create the local B tiles where A(k,k) is located
                 if (A.tileIsLocal(k, k)) {
+                    // XXX insert only what is needed for this iteration, otherwise,
+                    // all missing tiles are inserted.
                     for (int64_t j = 0; j < nt; ++j) {
-                        if (! B.tileIsLocal(k, j) && ! B.tileExists(k, j)) {
-                            B.tileInsert(k, j);
-                            B.at(k, j).set(0, 0);
+                        if (! B.tileIsLocal(k, j) ) {
+                            if (target == Target::Devices) {
+                                int device = A.tileDevice( k, k );
+                                if (! B.tileExists( k, j, device )) {
+                                    B.tileInsertWorkspace( k, j, device );
+                                    B.tileModified( k, j, device );
+
+                                    // XXX maybe reset the memory in case it got created from a previous call.
+                                    blas::Queue* queue = A.compute_queue( device, queue_0 );
+                                    assert( queue != nullptr );
+                                    auto T = B( k, j, device );
+                                    if (T.op() == Op::Trans || T.op() == Op::ConjTrans)
+                                        T = transpose( T );
+
+                                    B.tileGetForWriting( k, j, device,
+                                            LayoutConvert( layout ) );
```

Comment:
Is it overkill to GetForWriting when the tile was just inserted & marked as modified 10 lines above?

Reply:
I understand your remark. But we have no choice as far as I understand it.
When we insert a `workspace` tile, its state is `invalid` (see [BaseMatrix](https://github.com/icl-utk-edu/slate/blob/master/include/slate/BaseMatrix.hh#L1456) that calls [MatrixStorage](https://github.com/icl-utk-edu/slate/blob/master/include/slate/internal/MatrixStorage.hh#L1161)).
Therefore we cannot get this tile using `tileGet` (see [BaseMatrix](https://github.com/icl-utk-edu/slate/blob/master/include/slate/BaseMatrix.hh#L2780)).
But if I force the state by calling `tileModified`, that sets inconditionally the state to modified (see [BaseMatrix](https://github.com/icl-utk-edu/slate/blob/master/include/slate/BaseMatrix.hh#L1724)), then I can call `tileGetForWriting` (see [BaseMatrix]9https://github.com/icl-utk-edu/slate/blob/master/include/slate/BaseMatrix.hh#L3055)).

Reply:
Add `layout` to tileInsert, then shouldn't need tileGetForWriting.

Reply:
Done.